### PR TITLE
[mypyc] Add tests for str.method and fstring

### PR DESCRIPTION
**Pull request by 97littleleaf11**

### CODE REVIEW

File: mypyc/test-data/run-strings.test [hunk 1]
```diff
@@ -178,3 +148,164 @@ def test_str_to_bool() -> None:
...
+    assert '{:^30}'.format('centered') == '           centered           '
+    assert '{:*^30}'.format('centered') == '***********centered***********' # use '*' as a fill char
+
+    # Replacing %+f, %-f, and % f and specifying a sign:
+    assert '{:+f}; {:+f}'.format(3.14, -3.14) == '+3.140000; -3.140000' # show it always
+    assert '{: f}; {: f}'.format(3.14, -3.14) == ' 3.140000; -3.140000' # show a space for positive numbers
+    assert '{:-f}; {:-f}'.format(3.14, -3.14) == '3.140000; -3.140000' # show only the minus -- same as '{:f}; {:f}'
+
+    # Replacing %x and %o and converting the value to different bases:
+    s = "int: {0:d};  hex: {0:x};  oct: {0:o};  bin: {0:b}".format(42) # format also supports binary numbers
```
Comment (JukkaL) [inline mypyc/test-data/run-strings.test:271]: Test also a negative integer. Test large integer (>2**64).

File: mypyc/test-data/run-strings.test [hunk 2]
```diff
@@ -178,3 +148,164 @@ def test_str_to_bool() -> None:
...
+    assert f'Hello {var!a}' == "Hello 'mypyc'"
+    # ascii() is equivalent to !a
+    assert f'Hello {ascii(var)}' == "Hello 'mypyc'"
+
+    tmp_str = """this
+    is a new line."""
+    assert f'Test: {tmp_str!a}' == "Test: 'this\\n    is a new line.'"
+
+    assert f'Hello {var!s}' == "Hello mypyc"
+    assert f'Hello {num!s}' == "Hello 20"
```
Comment (JukkaL) [inline mypyc/test-data/run-strings.test:176]: Test an integer without `!s` (just `{}`). Test a negative integer. The big integer (>2**64), and also negative big integer.

Test `{}` with boolean, bytes and float arguments.

File: mypyc/test-data/run-strings.test [hunk 3]
```diff
@@ -178,3 +148,164 @@ def test_str_to_bool() -> None:
...
+        for base in 'dXob':
+            tmp_str += ('{0:{width}{base}}'.format(num, base=base, width=width))
+        tmp_strs.append(tmp_str)
+    assert tmp_strs == ['    5    5    5  101',\
+                        '    6    6    6  110',\
+                        '    7    7    7  111',\
+                        '    8    8   10 1000',\
+                        '    9    9   11 1001',\
+                        '   10    A   12 1010',\
+                        '   11    B   13 1011']
```
Comment (JukkaL) [inline mypyc/test-data/run-strings.test:311]: Test `{}` with bool, bytes and float arguments. Test both positive and negative floats + zero. Test a very large float and a float that is close to zero (e.g. 1.23e-20). Test infinities and NaN.

File: mypyc/test-data/run-strings.test [hunk 4]
```diff
@@ -178,3 +148,164 @@ def test_str_to_bool() -> None:
     for x in 'a', 'foo', 'bar', 'some string':
         assert is_true(x)
         assert not is_false(x)
+
+[case testFormattedStringLiterals]
```
Comment (JukkaL) [inline mypyc/test-data/run-strings.test:152]: Nit: Maybe call this `testFStrings`, since that's the most commonly used name for the feature.

**Review by JukkaL — COMMENTED** [review]
Great, having tests makes it easier to validate various optimizations. Looks good, just left some ideas about additional things to test.

File: mypyc/test-data/run-strings.test [hunk 5]
```diff
@@ -178,3 +148,164 @@ def test_str_to_bool() -> None:
...
+def test_fstring_simple() -> None:
+    assert f'Hello {var}, this is a test' == "Hello mypyc, this is a test"
+
+def test_fstring_conversion() -> None:
+    assert f'Hello {var!r}' == "Hello 'mypyc'"
+    # repr() is equivalent to !r
+    assert f'Hello {repr(var)}' == "Hello 'mypyc'"
+
+    assert f'Hello {var!a}' == "Hello 'mypyc'"
+    # ascii() is equivalent to !a
```
Comment (JelleZijlstra) [inline mypyc/test-data/run-strings.test:168]: Testing a string with non-ASCII characters could also be interesting here.

**Review by JukkaL — APPROVED** [review]
Thanks for the updates! Looks good now. This adds a good amount of test coverage that gives us confidence that any optimizations preserve existing behavior.

If you have any further test improvements in mind, they can be made in separate PRs.

**Pull request merged**